### PR TITLE
feature/return LGBMTunerCV booster

### PR DESCRIPTION
**Pull request by nyanhi**
## Motivation
To address #1609 

## Description of the changes
Add feature LightGBMTunerCV to return best booster
Please note that the test `lightgbm_tuner_test.test_get_best_booster` is intentionally omitting test for the argument `return_cvboost` due to version for now

### CODE REVIEW

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 1]
```diff
@@ -929,6 +942,8 @@ def __init__(
         optuna_callbacks: Optional[List[Callable[[Study, FrozenTrial], None]]] = None,
         verbosity: Optional[int] = None,
         show_progress_bar: bool = True,
+        model_dir: Optional[str] = None,
+        return_cvboost: Optional[bool] = None,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:946]: The argument name seems to be `return_cvbooster` according to the [reference](https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.cv.html).
```suggestion
        return_cvbooster: Optional[bool] = None,
```

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 2]
```diff
@@ -306,8 +306,16 @@ def __call__(self, trial: optuna.trial.Trial) -> float:
         elapsed_secs = time.time() - start_time
         average_iteration_time = elapsed_secs / len(val_scores)
 
+        if self.model_dir is not None and self.lgbm_params.get("return_cvboost"):
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:309]: ditto.
```suggestion
        if self.model_dir is not None and self.lgbm_params.get("return_cvbooster"):
```

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 3]
```diff
@@ -306,8 +306,16 @@ def __call__(self, trial: optuna.trial.Trial) -> float:
...
+            path = os.path.join(self.model_dir, "{}.pkl".format(trial.number))
+            with open(path, "wb") as fout:
+                pickle.dump(cv_results["cvbooster"], fout)
+            _logger.info("The booster of trial#{} was saved as {}.".format(trial.number, path))
+
         if self.compare_validation_metrics(val_score, self.best_score):
             self.best_score = val_score
+            if self.lgbm_params.get("return_cvboost"):
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:317]: ditto.
```suggestion
            if self.lgbm_params.get("return_cvbooster"):
```

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 4]
```diff
@@ -424,6 +435,46 @@ def best_params(self) -> Dict[str, Any]:
             params.update(self.lgbm_params)
             return params
 
+    def get_best_booster(self) -> "lgb.Booster":
+        """Return the best booster.
+
+        If the best booster cannot be found, :class:`ValueError` will be raised. To prevent the
+        errors, please save boosters by specifying either the ``model_dir`` argument of
+        :meth:`~optuna.integration.lightgbm.LightGBMTuner.__init__`,
+        or both of the ``model_dir`` and the ``return_cvboost`` arguments of
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:444]: ditto.
```suggestion
        or both of the ``model_dir`` and the ``return_cvbooster`` arguments of
```

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 5]
```diff
@@ -957,6 +973,8 @@ def __init__(
         self.lgbm_kwargs["show_stdv"] = show_stdv
         self.lgbm_kwargs["seed"] = seed
         self.lgbm_kwargs["fpreproc"] = fpreproc
+        if return_cvboost is not None:
+            self.lgbm_kwargs["return_cvboost"] = return_cvboost
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:977]: ditto.
```suggestion
        if return_cvbooster is not None:
            self.lgbm_kwargs["return_cvbooster"] = return_cvbooster
```

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 6]
```diff
@@ -949,3 +949,26 @@ def test_optuna_callback(self) -> None:
...
         assert callback_mock.call_count == 10
+
+    def test_get_best_booster(self) -> None:
+        unexpected_value = 20  # out of scope.
+
+        params = {"verbose": -1, "lambda_l1": unexpected_value}  # type: Dict
+        dataset = lgb.Dataset(np.zeros((10, 10)))
+
+        study = optuna.create_study()
+        # test case with the argument of `return_cvboost` is omitted as LGBM requires ver 3.0
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:960]: We usually put periods at the end of the sentences.
```suggestion
        # Test case with the argument of `return_cvbooster` is omitted as it requires `lightgbm>=3.0.0`.
```

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 2 above)
Comment (toshihikoyanase) [reply optuna/integration/_lightgbm_tuner/optimize.py:309]: In addition, I found that `self.model_dir` is always `None`. Could you add `model_dir` argument to `OptunaObjectiveCV.__init__` like `OptunaObjective.__init__`?

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your PR. I added some comments to save the models as expected.

**Comment by rafmacalaba** [comment]
@toshihikoyanase glad I saw this first before developing. this is my idea on how to optimize on feval on #1636, after merging this I will make it :D

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 7]
```diff
@@ -949,3 +949,26 @@ def test_optuna_callback(self) -> None:
...
         assert callback_mock.call_count == 10
+
+    def test_get_best_booster(self) -> None:
+        unexpected_value = 20  # out of scope.
+
+        params = {"verbose": -1, "lambda_l1": unexpected_value}  # type: Dict
+        dataset = lgb.Dataset(np.zeros((10, 10)))
+
+        study = optuna.create_study()
+        # Test with the argument of `return_cvbooster` is omitted as it requires `lightgbm>=3.0.0`.
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:960]: `LightGBM==3.0.0` has been released, and the known issue of  `LightGBMTuner` about `feature_pre_filter` was resolved by https://github.com/optuna/optuna/pull/1774. I think we can test the `return_cvbooster` argument.

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py (hunk 7 above)
Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Sure, noted your points. I will try to make the change tomorrow, thank you so much for your help as always.

Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: I think the test for `return_cvbooster` argument is added and the latest commit `ec1dba9` has passed mosts of tests. However only `ci/circleci: tests-python35` failed due to an ImportError for `skorch` below. 


```
==================================== ERRORS ====================================
___________ ERROR collecting tests/integration_tests/test_skorch.py ____________
ImportError while importing test module '/home/circleci/project/tests/integration_tests/test_skorch.py'.
Hint: make sure your test modules/packages have valid Python names.
Traceback:
/usr/local/lib/python3.5/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/integration_tests/test_skorch.py:3: in <module>
    import skorch
E   ImportError: No module named 'skorch'
```

Comment (toshihikoyanase) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Could you merge `master` into this feature branch again? https://github.com/optuna/optuna/pull/1668 introduced the change in `.circleci/config.yml` to ignore the `skorch`-related tests, but it may not be applied to this branch yet.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 8]
```diff
@@ -968,6 +986,8 @@ def __init__(
         self.lgbm_kwargs["show_stdv"] = show_stdv
         self.lgbm_kwargs["seed"] = seed
         self.lgbm_kwargs["fpreproc"] = fpreproc
+        if return_cvbooster is not None:
+            self.lgbm_kwargs["return_cvbooster"] = return_cvbooster
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:990]: Could you check the version of `lightgbm` to notify this feature is only for `lightgbm>=3.0.0`?
If users use `lightgbm<3.0.0`, they will get the following error and I think it is not user-friendly.

```console
  File "/Users/yanase/pfn/code/optuna/optuna/study.py", line 778, in _run_trial
    result = func(trial)
  File "/Users/yanase/pfn/code/optuna/optuna/integration/_lightgbm_tuner/optimize.py", line 305, in __call__
    cv_results = lgb.cv(self.lgbm_params, self.train_set, **self.lgbm_kwargs)
TypeError: cv() got an unexpected keyword argument 'return_cvbooster'
```

This is just and example. Please feel free to update it. Please note that the following code requires `packaging.version`.
```suggestion
        if return_cvbooster is not None:
            if version.parse(lgb.__version__) < version.parse("3.0.0"):
                raise ValueError("return_cvbooster requires lightgbm>=3.0.0.")
            self.lgbm_kwargs["return_cvbooster"] = return_cvbooster
```

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 9]
```diff
@@ -923,3 +923,44 @@ def test_optuna_callback(self) -> None:
...
+
+            best_booster = tuner.get_best_booster()
+            for booster in best_booster.boosters:
+                assert booster.params["lambda_l1"] != unexpected_value
+
+        tuner2 = LightGBMTuner(params, dataset, valid_sets=dataset, study=study)
+
+        # Resumed study does not have the best booster.
+        with pytest.raises(ValueError):
+            tuner2.get_best_booster()
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:953]: This test case is not related to `LightGBMTunerCV`. Could you remove it?
BTW, I think we can get the best booster if we specify `model_dir`. What do you think?

```python
        with TemporaryDirectory() as tmpdir:
            ...
            tuner2 = LightGBMTunerCV(params, dataset, study=study, model_dir=tmpdir, return_cvbooster=True)
            best_booster = tuner2.get_best_booster()
            for booster in best_booster.boosters:
                assert booster.params["lambda_l1"] != unexpected_value
```

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 10]
```diff
@@ -923,3 +923,44 @@ def test_optuna_callback(self) -> None:
...
+
+        # the `model_dir` argument is None
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:955]: Could you add a period at the end of the sentence?
```suggestion
        # The `model_dir` argument is None.
```

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 11]
```diff
@@ -923,3 +923,44 @@ def test_optuna_callback(self) -> None:
...
+        with pytest.raises(ValueError) as excinfo:
+            tuner3 = LightGBMTunerCV(
+                params,
+                dataset,
+                study=study,
+                model_dir=None,
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:961]: I think the case where `model_dir` is `None` and `return_cvbooster=True` is missing. I'm not sure if we can skip it in the test. Do you think it is redundant?

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
I added some comments on the test cases.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py (hunk 11 above)
Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Apologies for the delay in response. I have added the case where `model_dir` is `None` and `return_cvbooster=True`.

I would like to get your advice for the opposite case where `model_dir` **is not** `None` and `return_cvbooster=False`.

Current behavior of `get_best_booster()` is as following:
1. if the best booster cannot be found in `model_dir`, `get_best_booster()` raise ValueError, which is expected.
2. if the best booster is found in `model_dir`, `get_best_booster()` still returns `lightgbm.CVBooster`.

Behavior 2 was bit counterintuitive to me, and I thought adding following check may simplify it.
Is this relating to your comment?
>BTW, I think we can get the best booster if we specify model_dir. What do you think?
```
        if self.lgbm_kwargs.get("return_cvbooster") is not True:
            raise ValueError(
                "LightGBMTunerCV requires `return_cvbooster=True` for method `get_best_booster()`."
            )
```

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 12]
```diff
@@ -984,5 +1007,54 @@ def _create_objective(
...
+        To prevent the errors, please save boosters by specifying
+        both of the ``model_dir`` and the ``return_cvbooster`` arguments of
+        :meth: `~optuna.integration.lightgbm.LightGBMTunerCV.__init__`
+        when you resume tuning or you run tuning in parallel.
+        """
+        if self.lgbm_kwargs.get("return_cvbooster") is not True:
+            raise ValueError(
+                "LightGBMTunerCV requires `return_cvbooster=True` for method `get_best_booster()`."
+            )
+        if self._best_booster_with_trial_number is not None:
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:1027]: I found that `self._best_booster_with_trial_number` was always `None` since it was not set at `self._tune_params`. Now that both `LightGBMTuner` and `LightGBMTunerCV` uses `self._best_booster_with_trial_number`, how about adding the following line to the end of `_LightGBMBaseTuner._tune_params`?

```python
        if objective.best_booster_with_trial_number is not None:
            self._best_booster_with_trial_number = objective.best_booster_with_trial_number
```

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update.
I noticed that `LightGBMTuner` can return the best booster even if it does not receive the `model_dir` option while `LightGBMTunerCV` cannot. Could you check my comment to make the behavior of `get_best_booster` consistent?

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 12 above)
Comment (nyanhi) [reply optuna/integration/_lightgbm_tuner/optimize.py:1013]: Thank you for the clarification and suggestion. I was misunderstanding how `_best_booster_with_trial_number` works, especially when `model_dir==None`.
Have added the line as you suggested, and updated test code as well.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 13]
```diff
@@ -833,23 +836,6 @@ def best_booster(self) -> "lgb.Booster":
...
-        step_name: str,
-    ) -> _OptunaObjective:
-
-        objective = super(LightGBMTuner, self)._tune_params(
-            target_param_names, n_trials, sampler, step_name
-        )
-
-        if objective.best_booster_with_trial_number is not None:
-            self._best_booster_with_trial_number = objective.best_booster_with_trial_number
-            self._best_params.update(self.best_params)
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:834]: I noticed that `self._best_params` is not used by any part of the code. We can see two other lines (L387 and L401), so could you remove them too?

**Review by toshihikoyanase — COMMENTED** [review]
Thank you for your update! The logic of `LightGBMTunerCV` looks good to me, but I have some comments to improve maintenanceability of it and its tests.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 14]
```diff
@@ -925,48 +925,59 @@ def test_optuna_callback(self) -> None:
...
                         return_cvbooster=return_cvbooster,
                     )
-                    with mock.patch.object(
-                        _OptunaObjectiveCV, "_get_cv_scores", return_value=[1.0]
-                    ):
-                        tuner2.tune_regularization_factors()
-                    tuner2.get_best_booster()
+                    best_boosters2 = tuner2.get_best_booster().boosters
+                    for booster, booster2 in zip(best_boosters, best_boosters2):
+                        assert booster.params == booster2.params
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: I'm sorry I think you mentioned earlier as well. (Hopefully) I finally understood the logic, so made test cases for `model_dir is None` which resumed study done by different tuner and failed to find the best booster.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 15]
```diff
@@ -925,48 +925,59 @@ def test_optuna_callback(self) -> None:
         assert callback_mock.call_count == 10
 
     @pytest.mark.parametrize(
-        "model_dir, return_cvbooster", [(True, True), (False, True), (True, False), (False, False)]
+        "return_cvbooster, model_dir", [(True, True), (True, False), (False, True), (False, False)]
     )
-    def test_get_best_booster(self, model_dir: Optional[str], return_cvbooster: bool) -> None:
+    def test_get_best_booster(self, return_cvbooster: bool, model_dir: Optional[str]) -> None:
```
[the remark this answers is not in the document]
Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Thank you for the suggestion, I think it's good idea.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py [hunk 16]
```diff
@@ -923,3 +923,62 @@ def test_optuna_callback(self) -> None:
...
+            best_boosters = tuner.get_best_booster().boosters
+            for booster in best_boosters:
+                assert booster.params["lambda_l1"] != unexpected_value
+
+                tuner2 = LightGBMTunerCV(
+                    params, dataset, study=study, model_dir=tmpdir, return_cvbooster=True
+                )
+                best_boosters2 = tuner2.get_best_booster().boosters
+                for booster, booster2 in zip(best_boosters, best_boosters2):
+                    assert booster.params == booster2.params
```
Comment (toshihikoyanase) [inline tests/integration_tests/lightgbm_tuner_tests/test_optimize.py:954]: I think these lines are over-indented.
```suggestion
            tuner2 = LightGBMTunerCV(
                params, dataset, study=study, model_dir=tmpdir, return_cvbooster=True
            )
            best_boosters2 = tuner2.get_best_booster().boosters
            for booster, booster2 in zip(best_boosters, best_boosters2):
                assert booster.params == booster2.params
```

**Review by toshihikoyanase — CHANGES_REQUESTED** [review]
Thank you for your update. It basically looks good to me except the indent of the test.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (toshihikoyanase) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Thank you for adding this test case and re-organizing the tests for `LightGBMTunerCV.get_best_booster`. The current code is much easier to understand than the previous one.

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 13 above)
Comment (toshihikoyanase) [reply optuna/integration/_lightgbm_tuner/optimize.py:834]: Thank you for fixing them in https://github.com/optuna/optuna/pull/1702/commits/de499ce53148ab6e4bea7cb15348e37bed7397f7.

File: tests/integration_tests/lightgbm_tuner_tests/test_optimize.py (hunk 16 above)
Comment (nyanhi) [reply tests/integration_tests/lightgbm_tuner_tests/test_optimize.py]: Thank you for pointing out, fixed those lines.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 17]
```diff
@@ -984,5 +989,54 @@ def _create_objective(
...
             pbar=pbar,
         )
+
+    def get_best_booster(self) -> "lgb.CVBooster":
+        """Return the best cvbooster.
+
+        If the best booster cannot be found, :class:`ValueError` will be raised.
+        To prevent the errors, please save boosters by specifying
+        both of the ``model_dir`` and the ``return_cvbooster`` arguments of
+        :meth: `~optuna.integration.lightgbm.LightGBMTunerCV.__init__`,
```
Comment (toshihikoyanase) [inline optuna/integration/_lightgbm_tuner/optimize.py:1002]: A nit, but the document is not rendered correctly.

image

```suggestion
        :meth:`~optuna.integration.lightgbm.LightGBMTunerCV.__init__`,
```

**Review by toshihikoyanase — APPROVED** [review]
Thank you for your update! LGTM. I found a small mistake, but I can work on it in the follow-up PR if you are busy.

@HideakiImamura Please merge after your review.

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 17 above)
Comment (nyanhi) [reply optuna/integration/_lightgbm_tuner/optimize.py:1002]: Thank you for letting me know, I will quickly fix this. Sorry I just learnt this notation is for the document.

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the great improvement of the lightgbm integration. The changes almost LGTM.

I would like to know why `LightGBMTuner.get_best_booster()` is moved into `_LightGBMBaseTuner.get_best_booster()`. The document of `_LightGBMBaseTuner.get_best_booster()` seems to be only for `LightGBMTuner`.

File: optuna/integration/_lightgbm_tuner/optimize.py [hunk 18]
```diff
@@ -370,7 +384,8 @@ def __init__(
         self._parse_args(*args, **kwargs)
         self._start_time = None  # type: Optional[float]
         self._optuna_callbacks = optuna_callbacks
-        self._best_params = {}
+        self._best_booster_with_trial_number = None  # type: Optional[Tuple[lgb.Booster, int]]
```
Comment (HideakiImamura) [inline optuna/integration/_lightgbm_tuner/optimize.py:387]:
```suggestion
        self._best_booster_with_trial_number = None  # type: Optional[Tuple[Union[lgb.Booster, lgb.CVBooster], int]]
```

**Review by HideakiImamura — COMMENTED** [review]
I have a minor comment.

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 18 above)
Comment (nyanhi) [reply optuna/integration/_lightgbm_tuner/optimize.py:387]: Thank you so much for pointing out. I will make this change.

**Comment by nyanhi** [comment]
@HideakiImamura Thank you for your comment! Please find below response per my understanding, and kindly let me know your thought.

Given `lightgbm.cv()` also became enabled to return lightgbm.CVBooster since `lightgbm >= 3.0.0`, I think it would be natural to have same method `get_best_booster()` for both `LightGBMTuner` and `LightGBMTunerCV`. Hence I have moved the method into the base class. 

I see your point for the doc string and couple of error messages which are only meant for `LightGBMTuner`.  
One option I think of is to create an abstract method, similar to `_LightGBMBaseTuner._create_objective()`.

File: optuna/integration/_lightgbm_tuner/optimize.py (hunk 18 above)
Comment (nyanhi) [reply optuna/integration/_lightgbm_tuner/optimize.py:387]: Thank you for pointing out, please let me make this change.

**Comment by nyanhi** [comment]
I noticed 4 checks out of 19s are somehow still pending. Would it be possible to get advise how to complete it?
Apologies if anything is pending from my end...

**Comment by HideakiImamura** [comment]
Sorry for the late review. The requirement of the CI is recently migrated from CircleCI to GitHub actions. If you merge the master branch, this PR will pass the CI.


I will review this PR immediately. Sorry again for my late review.

**Review by HideakiImamura — APPROVED** [review]
Sorry for the late review. Thanks for the update!  LGTM!

**Comment by HideakiImamura** [comment]
@nyanhi Thank you for the update and remind. Could you merge the master branch? Then, the CI jobs in GitHub Actions will start.

**Comment by nyanhi** [comment]
@HideakiImamura Thank you for the review during busy time, much appreciated!! 
I will merge the mater branch this evening.

**Comment by HideakiImamura** [comment]
@nyanhi Thank you for merging the master branch. Let me merge this PR.
